### DISCUSSION TITLE
Running in to problems adding items to the inventory (InventoryItem)

### DISCUSSION THREAD

**Discussion by z3dd4-de** [peter-kish/gloot#137] [General]
Hi Peter!
I'm relatively new to Godot and just found GLoot a few days ago. In my game I have several items that can be dropped and the player should pick them up. I have a helper method for this inside player.gd - the player scene has an InventoryStacked and a CtrlInventoryStacked:

func addItem(item_name: String):
	var item = InventoryItem.new()
	item.prototype_id = item_name
	if inventory_stacked.has_place_for(item):
		inventory_stacked.add_item_automerge(item)

But it seems that this works only for the first item from the protoset which is this:
[
    {
        "id": "apple",
        "image": "res://art/apple-icon.png",
        "name": "Apple",
        "stack_size": 1
    },
    {
        "id": "stick",
        "image": "res://art/stick-icon.png",
        "name": "Stick",
        "stack_size": 5
    },
    {
        "id": "water_potion",
        "image": "res://art/water_potion.png",
        "name": "Water potion",
        "stack_size": 1
    },
    {
        "id": "slime",
        "image": "res://art/SlimeIcon.png",
        "name": "Slime",
        "stack_size": 1
    }
]

Whenever I add another item, I only get apples. When I remove the apple, I only get sticks (again the first item), but the stack_size = 5 works also only the first time. Then the default "1" will be added. Maybe I'm doing something completely wrong. The chance is very high that I maybe don't understand how to use your tool.
Take care.
z3dd4

**Comment by peter-kish** [comment]
I looked into it and I'm pretty sure this happens because the items created inside `addItem` are missing a protoset. Setting the protoset before the prototype ID should fix the issue:
```gdscript
func addItem(item_name: String):
    var item = InventoryItem.new()
    item.protoset = protoset
    item.prototype_id = item_name
    if inventory_stacked.has_place_for(item):
        inventory_stacked.add_item_automerge(item)
```

This can be considered a bug in the plugin which prevents you from setting the prototype ID while the protoset is `null`.

**Reply by z3dd4-de** [reply]
YES!!!!
Thank you so much, that was the solution. The protoset was null. Now it's working as expected and the stick has got the correct stack_size. For me it's fine now, but if I reported a bug, I'm happy that I could help to make your addon better.
Have a nice Christmas! 

```
func collect_item(item_name: String):
	var newitem = InventoryItem.new()
	newitem.protoset = inventory_stacked.item_protoset
	newitem.prototype_id = item_name
	if inventory_stacked.can_add_item(newitem):
		inventory_stacked.add_item_automerge(newitem)
```

**Reply by z3dd4-de** [reply]
Just as a follow-up:
In my game the player collects e.g. "apple"s. It's now working that they stack up according to the prototype's stack size in InventoryStacked: 1, 2, 3,...
Now I wanted to write a method "EatApple()" which removes ONE Apple from the stack.
But 
`inventory_stacked.remove_item(inventory_stacked.get_item_by_id("apple"))`
unfortunately removes ALL apples from the inventory.

I'm doing now the following which works, but is maybe a bit too complicated and not the obvious solution:
```
var size = inventory_stacked.get_item_by_id("apple").get_property("stack_size")
	print(size)
	if size >= 1:
		inventory_stacked.set_item_stack_size(inventory_stacked.get_item_by_id("apple"), size - inventory_stacked.get_prototype_stack_size("apple"))
```
I'm doing this in code, because it is a KeyPressed input to eat the apple and not depending whether the inventory is visible or not.

**Reply by peter-kish** [reply]
Maybe this would be a cleaner way to do it:
```gdscript
var item = inventory_stacked.get_item_by_id("apple")
inventory_stacked.set_item_stack_size(item, inventory_stacked.get_item_stack_size(item) - 1)
```
Note that once the item stack size reaches 0, the item stack should be removed from the inventory and queued for deletion.

**Reply by z3dd4-de** [reply]
That looks good. I haven't had the time to look into this over Christmas, but I will check it out and give you feedback. Thanks for your answer!
